### PR TITLE
Start emitting labels even if their pointed to file is not available locally

### LINKED ISSUE #97699: const_err spans not rendered when rust-src not installed
This code causes a const_err somewhere in the standard library:
```rust
#![feature(const_swap)]
#![feature(const_mut_refs)]
use std::{
    mem::{self, MaybeUninit},
    ptr,
};

const X: () = {
    let mut ptr1 = &1;
    let mut ptr2 = &2;

    // Swap them, bytewise.
    unsafe {
        ptr::swap_nonoverlapping(
            &mut ptr1 as *mut _ as *mut MaybeUninit<u8>,
            &mut ptr2 as *mut _ as *mut MaybeUninit<u8>,
            mem::size_of::<&i32>(),
        );
    }
};
```

When the `rust-src` package is not installed, the error is pretty unreadable:
```
error: any use of this value will cause an error
   |
  ::: src/lib.rs:8:1
   |
8  | / const X: () = {
9  | |     let mut ptr1 = &1;
10 | |     let mut ptr2 = &2;
11 | |
...  |
19 | |     }
20 | | };
   | |__-
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, [see issue #71800 <https://github.com/rust-lang/rust/issues/71800>](https://github.com/rust-lang/rust/issues/71800)
```

This is missing the key information of what goes wrong! With `rust-src`, the following is added, making it much clearer:
```
    --> /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:1151:9
     |
1151 |           copy_nonoverlapping(src, tmp.as_mut_ptr(), 1);
     |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |           |
     |           unable to turn pointer into raw bytes
     |           inside `std::ptr::read::<MaybeUninit<u8>>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:1151:9
     |           inside `mem::swap_simple::<MaybeUninit<u8>>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:768:17
     |           inside `ptr::swap_nonoverlapping_simple::<MaybeUninit<u8>>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:968:9
     |           inside `swap_nonoverlapping::<MaybeUninit<u8>>` at /home/r/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:949:14
     |           inside `X` at test.rs:14:9
```

If the sources are missing, we should probably still show the error, even if we cannot attach it to any code.
Cc @rust-lang/wg-const-eval

### PR DESCRIPTION
**Pull request by oli-obk**
r? @estebank

cc @RalfJung 

fixes #97699

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -256,25 +256,13 @@ impl<'tcx> fmt::Display for FrameInfo<'tcx> {
...
-                let lo = sm.lookup_char_pos(self.span.lo());
-                write!(
-                    f,
-                    " at {}:{}:{}",
-                    sm.filename_for_diagnostics(&lo.file.name),
-                    lo.line,
-                    lo.col.to_usize() + 1
-                )?;
-            }
-            Ok(())
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:277]: Ah I guess this is redundant... makes sense.

**Review by RalfJung — APPROVED** [review]
Can't check the `emitter` stuff, but the CTFE part looks good!

File: compiler/rustc_const_eval/src/interpret/eval_context.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:277]: in miri it isn't so... maybe I need to find a better way

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:277]: (just force pushed the last commit if you wanna have a look)

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:277]: Ah right, in Miri we don't always emit the span

https://github.com/rust-lang/rust/blob/2816fecd26eaa5cb8d3e0070d2d924b1ed284220/src/tools/miri/src/diagnostics.rs#L355-L359

but it probably makes more sense to make the code that decides not to print the span, compensate for that by adding location information to what it prints?

File: src/tools/miri/src/diagnostics.rs [hunk 2]
```diff
@@ -359,12 +359,14 @@ fn report_msg<'tcx>(
     }
     // Add backtrace
     for (idx, frame_info) in stacktrace.iter().enumerate() {
-        let is_local = machine.is_local(frame_info);
         // No span for non-local frames and the first frame (which is the error site).
-        if is_local && idx > 0 {
+        if idx > 0 {
```
Comment (saethlin) [inline src/tools/miri/src/diagnostics.rs:363]: Removing this `is_local` check blows up the size of Miri backtraces with information which is not useful to the user. Is there a principled reason we can't have this check anymore?

Comment (oli-obk) [reply src/tools/miri/src/diagnostics.rs:363]: nope ^^ I just forgot about that

File: src/test/ui/infinite/infinite-recursion-const-fn.stderr [hunk 3]
```diff
@@ -5,140 +5,140 @@ LL |     b()
...
-   |     inside `b` at $DIR/infinite-recursion-const-fn.rs:7:5
-   |     inside `b` at $DIR/infinite-recursion-const-fn.rs:7:5
-   |     inside `b` at $DIR/infinite-recursion-const-fn.rs:7:5
-   |     inside `b` at $DIR/infinite-recursion-const-fn.rs:7:5
-   |     inside `b` at $DIR/infinite-recursion-const-fn.rs:7:5
+   |     inside `b`
+   |     inside `b`
+   |     inside `b`
+   |     inside `b`
+   |     inside `b`
```
Comment (estebank) [inline src/test/ui/infinite/infinite-recursion-const-fn.stderr:80]: It seems like we could afford to add some deduplication here (probably best on its own PR)

File: src/test/ui/consts/recursive.stderr [hunk 4]
```diff
@@ -16,11 +16,11 @@ LL |     f(x);
    |     ^^^^
    |     |
    |     reached the configured maximum number of stack frames
-   |     inside `f::<i32>` at $DIR/recursive.rs:4:5
-   |     [... 126 additional calls inside `f::<i32>` at $DIR/recursive.rs:4:5 ...]
+   |     inside `f::<i32>`
+   |     [... 126 additional calls inside `f::<i32>` ...]
```
Comment (estebank) [inline src/test/ui/consts/recursive.stderr:20]: Oh! We already do some deduplication!

File: src/test/ui/consts/missing_span_in_backtrace.stderr [hunk 5]
```diff
@@ -0,0 +1,23 @@
+error[E0080]: evaluation of constant value failed
+/rustc/xyz/library/core/src/ptr/mod.rs:925:14: inside `swap_nonoverlapping::<MaybeUninit<u8>>`
+/rustc/xyz/library/core/src/ptr/mod.rs:944:9: inside `ptr::swap_nonoverlapping_simple_untyped::<MaybeUninit<u8>>`
+--> /rustc/xyz/library/core/src/ptr/mod.rs:1135:9
+note: unable to copy parts of a pointer from memory at alloc10
+note: inside `std::ptr::read::<MaybeUninit<MaybeUninit<u8>>>`
+/rustc/xyz/library/core/src/mem/mod.rs:773:17: inside `mem::swap_simple::<MaybeUninit<MaybeUninit<u8>>>`
+   |
+  ::: $DIR/missing_span_in_backtrace.rs:16:9
```
Comment (estebank) [inline src/test/ui/consts/missing_span_in_backtrace.stderr:9]: I think the expected output for these (which doesn't *need* to be followed, but I'd prefer to) would be

```suggestion
error[E0080]: evaluation of constant value failed
  |
  = /rustc/xyz/library/core/src/ptr/mod.rs:925:14: inside `swap_nonoverlapping::<MaybeUninit<u8>>`
  = /rustc/xyz/library/core/src/ptr/mod.rs:944:9: inside `ptr::swap_nonoverlapping_simple_untyped::<MaybeUninit<u8>>`
 --> /rustc/xyz/library/core/src/ptr/mod.rs:1135:9
  |
  = note: unable to copy parts of a pointer from memory at alloc10
  = note: inside `std::ptr::read::<MaybeUninit<MaybeUninit<u8>>>`
  = /rustc/xyz/library/core/src/mem/mod.rs:773:17: inside `mem::swap_simple::<MaybeUninit<MaybeUninit<u8>>>`
   |
  ::: $DIR/missing_span_in_backtrace.rs:16:9
```

Aligning with the available path (so accounting for the biggest number shown in the gutter) and separating the message from the notes given no available span.

File: compiler/rustc_errors/src/emitter.rs [hunk 6]
```diff
@@ -1410,6 +1406,42 @@ impl EmitterWriter {
         for annotated_file in annotated_files {
             // we can't annotate anything if the source is unavailable.
             if !sm.ensure_source_file_source_present(annotated_file.file.clone()) {
+                if !self.short_message {
+                    // We'll just print an unannotated message.
```
Comment (estebank) [inline compiler/rustc_errors/src/emitter.rs:1410]: Here is where we'd have to duplicate some of the printing logic for accurate gutter alignment. :-/

Ideally we'd dedup the logic, but it is fine as long as we have test coverage (which we do) and leave a comment in both places reminding us to modify logic in the other.

File: src/test/ui/consts/missing_span_in_backtrace.stderr (hunk 5 above)
Comment (estebank) [reply src/test/ui/consts/missing_span_in_backtrace.stderr:9]: Also, what about the following formatting?

```suggestion
error[E0080]: evaluation of constant value failed
  --> /rustc/xyz/library/core/src/ptr/mod.rs:1135:9
   |
note: inside `swap_nonoverlapping::<MaybeUninit<u8>>`
  --> /rustc/xyz/library/core/src/ptr/mod.rs:925:14
   |
note: inside `ptr::swap_nonoverlapping_simple_untyped::<MaybeUninit<u8>>`
  --> /rustc/xyz/library/core/src/ptr/mod.rs:944:9
   |
   = note: unable to copy parts of a pointer from memory at alloc10
   = note: inside `std::ptr::read::<MaybeUninit<MaybeUninit<u8>>>`
   = /rustc/xyz/library/core/src/mem/mod.rs:773:17: inside `mem::swap_simple::<MaybeUninit<MaybeUninit<u8>>>`
   |
  ::: $DIR/missing_span_in_backtrace.rs:16:9
```

I also noticed that the new pseudo-notes come before the path for the main diagnostic, that should be changed.

**Review by estebank — APPROVED** [review]
I'm +1 on the changes as they are, but I have a bunch of nitpicks on the resulting output. The one I would love to have addressed before merging is ensuring the `--> main diagnostic path` line comes before all the recovered span labels.

File: src/test/ui/consts/missing_span_in_backtrace.stderr (hunk 5 above)
Comment (estebank) [reply src/test/ui/consts/missing_span_in_backtrace.stderr:9]: One last thing: if there are span labels for the primary span, they should be displayed first and as a `= note:` without a path. I think that with all of these we stay inside the established grammar of the printed diagnostics.

Comment (oli-obk) [reply src/test/ui/consts/missing_span_in_backtrace.stderr:9]: At the emitter stage we do not have enough information to figure this out. Labels are reordered to fit the file structure. I will check if converting them to notes at const eval error time will still yield reasonable diagnostics

**Review by estebank — APPROVED** [review]

File: src/tools/miri/src/diagnostics.rs [hunk 7]
```diff
@@ -362,7 +362,10 @@ fn report_msg<'tcx>(
         if is_local && idx > 0 {
             err.span_note(frame_info.span, &frame_info.to_string());
         } else {
-            err.note(&frame_info.to_string());
+            let sm = sess.source_map();
+            let lo = sm.lookup_char_pos(frame_info.span.lo());
+            let filename = sm.filename_for_diagnostics(&lo.file.name);
+            err.note(format!("{frame_info} at {}:{}:{}", filename, lo.line, lo.col.0 + 1));
```
Comment (RalfJung) [inline src/tools/miri/src/diagnostics.rs:368]: Isn't there a utility function somewhere that renders a span as a string? Doesn't feel great to have that logic here, in particular with that `+1`.

Comment (RalfJung) [reply src/tools/miri/src/diagnostics.rs:368]: Now that we only print this when we don't have a span to show, maybe using `span_to_string` and showing the end makes sense?

Comment (oli-obk) [reply src/tools/miri/src/diagnostics.rs]: no diff in CI because we filter too heavily in the miri test suite

**Comment by oli-obk** [comment]
@bors retry [rustdoc] src/test/rustdoc/primitive/no_std.rs' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }

I ran rustdoc tests locally and they passed, so guessin spurious file system issue

**Comment by ehuss** [comment]
@bors retry

Failed to update toolstate
(has been fixed)

**Comment by RalfJung** [comment]
Is it only this PR that keeps failing with coverage reports? That would be a bit suspicious.
@bors retry

**Comment by RalfJung** [comment]
I can't even find an issue tracking that `clear_expected_if_blessed` is noisy?

**Comment by ehuss** [comment]
@RalfJung It's all PRs.  I started a thread on #t-infra.  I'm tempted to close the tree, but it is not a 100% failure.

**Pull request merged**